state/txindex/kv/kv.go

```diff
@@ -40,6 +41,21 @@ type TxIndex struct {
 	log log.Logger
 }
 
+func (txi *TxIndex) Prune(retainHeight int64) {
+	ctx := context.Background()
+	results, err := txi.Search(ctx, query.MustCompile(fmt.Sprintf("tx.height < %d", retainHeight)))
```

Comment:
Same comment as for the block indexer. Can we not just scan the DB and try to delete a height. If it is not there, it's not there. This way we do a search to find the heights that are there and then we do a scan anyways and try to delete those keys.

Reply:
To me, it is not as easy as with blocks, since here `Search` returns the results, not just the heights. In particular, I can optimize the deletion of events (see lines 217-228):
```[go]
zeroKey := keyForEvent(compositeTag, attr.Value, result, 0)
endKey := keyForEvent(compositeTag, attr.Value, result, math.MaxInt64)
itr, err := txi.store.Iterator(zeroKey, endKey)
if err != nil {
	panic(err)
}
for ; itr.Valid(); itr.Next() {
	err := txi.store.Delete(itr.Key())
	if err != nil {
		panic(err)
	}
}
```

Here I use the `result` found by the search not to traverse all the keys, but only those that are for specific height and event type